### PR TITLE
5010 – Fix tag with trailing space error

### PR DESCRIPTION
**Pull request by vasconsaurus**
## Description

When trying to create an imported fact-check that had a tag with a trailing space it would fail (`text has already been taken`).

We have a normalization step but that runs `before_validation` on `TagText`.
So when we are checking if the tag exists in line 92 from `tag.rb`, the tag has not been normalized. This means:

- We look for 'tag', it is not there. We create it.
- We look for ' tag', it is not there. We try to create it, it gets normalized, so it becomes 'tag'. Then we fail to create it,
because it already exists.

Adding the strip when comparing tags fixes this, but I'm not sure if it is the best solution.

References: 5010

## How has this been tested?

- I added one big test that replicated the issue we saw: `rails test test/controllers/graphql_controller_12_test.rb:613`
- And added one model test: `rails test test/models/tag_test.rb:275`

## Things to pay attention to during code review

Please describe parts of the change that require extra attention during code review, for example:

- File FFFF, line LL: This refactoring does this and this. Is it consistent with how it’s implemented elsewhere?
- Etc.

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

### CODE REVIEW

**Comment by vasconsaurus** [comment]
@caiosba Could you please take a look?
- This fixes it, but I'm not sure if it's a good solution.
- And although the first test helped me get this to work, I'm not sure we should keep it. WDYT?

File: app/models/annotations/tag.rb [hunk 1]
```diff
@@ -89,7 +89,7 @@ def self.run_bulk_create_callbacks(ids_json, pmids_json)
   def get_tag_text_reference
     if self.tag.is_a?(String)
       team_id = self.team&.id
-      tag_text = TagText.where(text: self.tag, team_id: team_id).last
+      tag_text = TagText.where(text: (self.tag).strip, team_id: team_id).last
```
Comment (caiosba) [inline app/models/annotations/tag.rb:92]: Minor, only FYI: The parenthesis are not needed here.

**Comment by caiosba** [comment]
That's a good solution, Manu. And thanks for adding the tests. Were you able to reproduce the same exception in the test, before the fix?

File: app/models/annotations/tag.rb [hunk 2]
```diff
@@ -89,7 +89,7 @@ def self.run_bulk_create_callbacks(ids_json, pmids_json)
   def get_tag_text_reference
     if self.tag.is_a?(String)
       team_id = self.team&.id
-      tag_text = TagText.where(text: self.tag, team_id: team_id).last
+      tag_text = TagText.where(text: self.tag.strip, team_id: team_id).last
```
Comment (melsawy) [inline app/models/annotations/tag.rb:92]: Also I suggest to trim database value so we can catch existing one with trailing space
i.e
tag_text = TagText.where(team_id: team_id).where('trim(text) = ?', self.tag.strip).last

Comment (melsawy) [reply app/models/annotations/tag.rb:92]: Oh I see that we normalize tag before saving in DB and did `self.text.strip.gsub(/^#/, '')` so we can skip this comment

**Review by melsawy — APPROVED** [review]

**Review by caiosba — APPROVED** [review]

**Pull request merged**